File: Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param head a ListNode
     * @return a boolean
     */
    bool isPalindrome(ListNode* head) {
        // Write your code here
        if (!head)
        {
            return true;
        }
        ListNode dummy(INT_MIN);
        dummy.next = head;
        ListNode *fast = &dummy;
        ListNode *slow = &dummy;
        while (fast->next && fast->next->next)
        {
            fast = fast->next->next;
            slow = slow->next;
        }
        slow = slow->next;
        ListNode *last = slow;
        ListNode *right_left = slow->next;
        while (right_left)
        {
            ListNode *tmp_right = right_left;
            right_left = right_left->next;
            tmp_right->next = last;
            last = tmp_right;
        }
        // return true;
        ListNode *first = head;
        while (first != last)
        {
            if (first->val != last->val)
            {
                return false;
            }
            if (first->next == last)
            {
                return true;
            }
            first = first->next;
            last = last->next;
        }
        return true;
    }
};

File: Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a boolean
    def isPalindrome(self, head):
        # Write your code here
        
        def reverse(root):
            cur = None
            while root:
                tmp = root.next
                root.next = cur
                cur = root
                root = tmp
            return cur
        
        if not head:
            return True
            
        dummy = ListNode(-1)
        dummy.next = head
        fast, slow = dummy, dummy
        while fast and fast.next:
            fast, slow = fast.next.next, slow.next
        
        nextHead = slow.next
        nextHead = reverse(nextHead)
        cur = nextHead
        ans = True
        while cur:
            if head.val == cur.val:
                head = head.next
                cur = cur.next
            else:
                ans = False
                break
        
        ## recover the data
        nextHead = reverse(nextHead)
        head = dummy.next
        slow.next = nextHead
        return ans